ChatSecure/Classes/Model/Yap Storage/OTRBaseMessage.m

```diff
@@ -251,6 +251,21 @@ + (void)deleteAllMessagesWithTransaction:(YapDatabaseReadWriteTransaction*)trans
 
 + (void)deleteAllMessagesForBuddyId:(NSString *)uniqueBuddyId transaction:(YapDatabaseReadWriteTransaction*)transaction
 {
+    //#865
+    NSArray <id <OTRMessageProtocol>>* mediaMessages = nil;
+    __block OTRMediaItem *item = nil;
+    OTRBuddy *buddy1 = [OTRBuddy fetchObjectWithUniqueID:uniqueBuddyId transaction:transaction];
+    id <OTRThreadOwner> threadOwner = buddy1;
+    mediaMessages = [transaction allMediaMessagesForThread:threadOwner];
+    [mediaMessages enumerateObjectsUsingBlock:^(id<OTRMessageProtocol>  _Nonnull medmsg, NSUInteger idx, BOOL * _Nonnull stop) {
+        item = [OTRMediaItem mediaItemForMessage:medmsg transaction:transaction];
+        [[OTRMediaFileManager sharedInstance] deleteDataForItem:item buddyUniqueId:uniqueBuddyId completion:^(NSError *error) {
+            if (error) {
+                DDLogError(@"ERROR in deleting data for media item");
+            }
+        } completionQueue:dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)];
+    }];
+    
```

Comment:
This can be improved by using YapDatabase deletion notify in both OTRBaseMessage and OTRXMPPRoomMessage, or perhaps OTRMediaItem: https://github.com/yapstudios/YapDatabase/wiki/Relationships#notify

```objc
- (nullable id)yapDatabaseRelationshipEdgeDeleted:(YapDatabaseRelationshipEdge *)edge
                                       withReason:(YDB_NotifyReason)reason {
// delete the data here
}
```

Normally Yap will automatically delete child objects (e.g. messages are deleted when a buddy is deleted) if the relationships are setup properly. In theory this can be chained down to the OTRMediaItem, and then `[[OTRMediaFileManager sharedInstance] deleteDataForItem` can be called inside `yapDatabaseRelationshipEdgeDeleted:withReason:` for each item individually.

Reply:
In this case, then the allMediaMessagesForThread question above is irrelevant and we don't need that function, right?

Reply:
In OTRBaseMessage.yapDatabaseRelationshipEdges, it appears to create the relationship for OTRMediaItems with nodeDeleteRules:YDB_DeleteDestinationIfSourceDeleted | YDB_NotifyIfSourceDeleted. 

But yapDatabaseRelationshipEdgeDeleted in OTRMediaItem doesn't get called when I delete a thread with media items. I'm not sure why. yapDatabaseRelationshipEdgeDeleted in OTRMediaItem doesn't call deleteDataForItem yet, but I just wanted to see if the method got called. It seems like it would be better to use yapDatabaseRelationshipEdgeDeleted in OTRMediaItem, but since I'm not sure why it isn't getting called, maybe I'll try in OTRBaseMessage and OTRXMPPRoomMessage

Reply:
The yapDatabaseRelationshipEdgeDeleted selector is never getting called from the flush method in YapDatabaseRelationshipTransaction. Was tracking down why, but had to do other things and ran out of time

Reply:
Thanks for your work here, I think you are pretty close and just need to play around with the nodeDeleteRules (perhaps you want destination not source?). You can repurpose some of the existing tests to speed things up if you'd like. If you see FileTransferIntegrationTests.swift you'll see an example of how to set up a test database. Just make sure when making your test data that you don't have an orphan object or it'll get deleted out from under you. (e.g. make Account, Buddy, and Message with proper relationships).. see FileTransferIntegrationTests.setUp()

Reply:
OTRBaseMessage currently has YDB_DeleteDestinationIfSourceDeleted, which deletes the media item (in YapDatabaseRelationshipTransaction) without sending notification. 

If we change it to YDB_NotifyIfSourceDeleted, it will send the notification and call yapDatabaseRelationshipEdgeDeleted in OTRMediaItem (where we could remove the file with deleteDataForItem), but it doesn't actually delete the destination edge item in this case. Returning nil does nothing, otherwise you can return a modified destination object which will update the database

So, I'm thinking there are a few possibilities, but maybe I'm missing something. It seems to me like YapDatabaseRelationshipTransaction should do one of the following:

1) Another nodeDeleteRule for YDB_DeleteDestinationAndNotifyIfSourceDeleted (same for source).
2) Have the option to delete if user returns nil from yapDatabaseRelationshipEdgeDeleted. However, this would mean any user would need to return the same, unmodified object if they DON'T want to delete in this case.

The flush method is already huge with all the conditions, but having the option to Notify and Delete would be helpful in this case.

All that being said, the current fix that I proposed in the original pull request is inefficient (as you said in your first comments), but it works. Sooooo...should we just hold off on the whole thing for now and just put a change request in for the YapDatabase pod?

Reply:
Oh interesting, so the "proper fix" is blocked on a YapDatabase bug? If that's the case, definitely open an issue on YapDatabase describing the problem we're seeing here, and I'll reach out to Robbie to see if he has any thoughts before we issue a pull request. 

I'd prefer to do it right the first time, because this will allow you to properly delete individual messages, which is something people have asked about for a long time.

Reply:
I just realized one more thing I need to test based on their example. I might be wrong and we might just need to change the rule. If that test fails, then I'll post on YapDatabase.

Reply:
My other test failed. Here's the ticket I submitted: https://github.com/yapstudios/YapDatabase/issues/426

Reply:
Thank you!